docs/design.md

```diff
@@ -0,0 +1,199 @@
+
+The file contains 2 value which represent the allowed time quota (microseconds) and the 
+period length (microseconds).
+
+## GRPC API
+
+The service exposes a gRPC API that gives capabilities to start a process, stop a process, get process status and stream process output.
+
+The proposed proto file can be found here: [worker_official.proto](./worker.proto)
+
+The GRPC server will be calling methods in the worker library to interact with processes.
+
+### Authentication
+
+The service will use mTLS to secure traffic and verify client and server authenticity. For this design document, we will make the assumption that the Certificate Authority (CA) can be trusted i.e. certificates are only issued to users and nodes that have been verified.
+
+Both server and client will support TLS versions 1.2 and 1.3.
+
+The [grpc-go](https://github.com/grpc/grpc-go) package that we will be using to implement the server and client
+give direct access to the standard library's `tls.Config` struct so it should be relatively simple to verify both
+client and server certificates.
+
+### Cipher Suites
+
+#### TLS 1.2
+
+Following the guidance from these RFCs:
+
+[Recommendations for Secure Use of Transport Layer Security and Datagram Transport Layer Security (DTLS)](https://www.ietf.org/archive/id/draft-ietf-uta-rfc7525bis-03.html)
+
+[Hypertext Transfer Protocol Version 2 (HTTP/2) Appendix A](https://datatracker.ietf.org/doc/html/rfc7540#appendix-A)
+
+We will only accept cipher suites that match the following criteria:
+* Key exchange algorithm supports perfect forward secrecy
+* No RC4 cipher suites
+* No cipher suites including algorithms offering less than 128 bits of security
+* No cipher suites which use static RSA i.e cipher suites starting with TLS_RSA_WITH_*
+* Cipher Suite is not blacklisted by HTTP/2
+
+This leaves us with the following cipher suites supported by the Go TLS package https://pkg.go.dev/crypto/tls :
+
+TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
+TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
+TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
+TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
+TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256
+TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256
+
+#### TLS 1.3
+
+TLS 1.3 offers enhanced security over its predecessor, all its supported cipher suites can be negotiated.
+
+### Authorisation
+
+For authorisation, a simple all-or-nothing mechanism will be used. The server can be started with a list of OS users that are allowed to make requests.
+
+X.509 certificates issued to users will have information about the user name encoded in them in the Common Name subject field. The server will then use this information to decide if a requester has been given permission to make requests.
+
+For example, using the snippet below the users `hashi`, `mozart` and `simba` would be able to make requests to the GRPC API.
+
+```
+./worker --allowedUsers=hashi,mozart,simba
+```
+
+## Worker Library
```

Comment:
Also in regards to the library, could you provide a brief explanation on how you intend to implement output streaming?

Reply:
Done!